macaroon.js

```diff
@@ -707,125 +845,168 @@ const Macaroon = class Macaroon {
+    return importFromJSONObjectV1(obj);
   }
+  return importFromJSONObjectV2(obj);
+};
+
+const importFromJSONObjectV1 = function(obj) {
+  const caveats = obj.caveats && obj.caveats.map(jsonCaveat => {
+    const caveat = {
+      identifierBytes: stringToBytes(requireString(jsonCaveat.cid, 'Caveat id')),
+      locationStr: maybeString(jsonCaveat.cl, 'Caveat location'),
+    };
+    if (jsonCaveat.vid) {
+      caveat.vidBytes = base64ToBytes(requireString(jsonCaveat.vid, 'Caveat verification id'));
+    }
+    return caveat;
+  });
   return new Macaroon({
-    signature: bytesToBitArray(hexToBytes(obj.signature)),
-    location: maybeString(obj.location, 'Macaroon location'),
-    identifier: requireString(obj.identifier, 'Macaroon identifier'),
+    version: 1,
+    locationStr: maybeString(obj.location, 'Macaroon location'),
+    identifierBytes: stringToBytes(requireString(obj.identifier, 'Macaroon identifier')),
     caveats: caveats,
+    signatureBytes: hexToBytes(obj.signature),
   });
 };
 
-
-function deserializeJsonField(obj, key, required) {
-  if (obj.hasOwnProperty(key + '64')) {
-    return sjcl.codec.base64.toBits(Buffer.from(obj[key + '64'], 'base64').toString('base64'));
-  } else if (obj.hasOwnProperty(key)) {
-    return obj[key];
-  } else if (required) {
-    throw new Error('Expected key: ' + key);
-  } else {
-    return null;
-  }
-}
-
 /**
  * Deserializes V2 JSON macaroon encoding.
  * @param {Object|Array} obj A serialized JSON macaroon
  * @return {Macaroon}
 */
-const deserializeJson = function(obj) {
+const importFromJSONObjectV2 = function(obj) {
   if (obj.v !== 2) {
-    throw new Error('Only version 2 is supported');
+    throw new Error(`Unsupported macaroon version ${obj.v}`);
   }
-  const caveats = !Array.isArray(obj.c) || obj.c.length === 0 ?
-    [] :
-    obj.c.map(caveat => {
+  const params = {
+    version: 2,
+    signatureBytes: v2JSONField(obj, 's', true),
+    locationStr: bytesToString(v2JSONField(obj, 'l', false)),
+    identifierBytes: v2JSONField(obj, 'i', true),
+  };
+  if (obj.c) {
+    if (!Array.isArray(obj.c)) {
+      throw new Error('caveats field does not hold an array');
+    };
+    params.caveats = obj.c.map(caveat => {
       return {
-        _identifier: deserializeJsonField(caveat, 'i', true),
-        _location: deserializeJsonField(caveat, 'l'),
-        _vid: deserializeJsonField(caveat, 'v')
+        identifierBytes: v2JSONField(caveat, 'i', true),
+        locationStr: bytesToString(v2JSONField(caveat, 'l'), false),
+        vidBytes: v2JSONField(caveat, 'v', false)
       };
     });
-  return new Macaroon({
-    signature: deserializeJsonField(obj, 's', true),
-    location: deserializeJsonField(obj, 'l'),
-    identifier: deserializeJsonField(obj, 'i', true),
-    caveats: caveats,
-  });
+  }
+  return new Macaroon(params);
+};
+
+/**
+ * Read a JSON field that might be in base64 or string
+ * format.
+ * @param {Object} obj A deserialized JSON object.
```

Comment:
You can, but don't need to, specify the keys and types in JSDoc:

```
@param {Object} obj
@param {string} obj.key1
```

Reply:
If I understand you correctly, I'm not sure that applies here. We don't know what type is in that field (it's come from some external source), and actually we don't even know what the field name is yet.

Reply:
:+1: